tests/integration/_experimental/dataset/test_evaluation.py

```diff
@@ -0,0 +1,306 @@
+    )
+
+    df_by_eval = fetch_evaluation_results(dataset_name, model_name)
+    eval_cfg, df_datapoints, df_inferences, df_metrics = df_by_eval[0]
+    expected_df_dp = df_dp[13:20].reset_index(drop=True)
+    expected_df_inf = pd.concat([df_inf[13:20]], axis=0).reset_index(
+        drop=True,
+    )
+    expected_df_mtr = pd.concat([df_mtr[13:20]], axis=0).reset_index(
+        drop=True,
+    )
+    assert len(df_by_eval) == 1
+    assert eval_cfg is None
+    _assert_frame_equal(df_datapoints, expected_df_dp, dp_columns)
+    _assert_frame_equal(df_inferences, expected_df_inf, inf_columns)
+    _assert_frame_equal(df_metrics, expected_df_mtr, mtr_columns)
+
+    # add 3 new datapoints, then we should have missing inference and metrics in the db records
+    register_dataset(dataset_name, df_dp[10:20][dp_columns])
+
+    df_by_eval = fetch_evaluation_results(dataset_name, model_name)
+    eval_cfg, df_datapoints, df_inferences, df_metrics = df_by_eval[0]
+    expected_df_dp = pd.concat([df_dp[13:20], df_dp[10:13]]).reset_index(drop=True)
+    expected_df_inf = pd.concat(
+        [df_inf[13:20], pd.DataFrame({"softmax_bitmap": [np.nan] * 3})],
+        axis=0,
+    ).reset_index(drop=True)
+    expected_df_mtr = pd.concat([df_mtr[13:20], pd.DataFrame({"score": [np.nan] * 3})], axis=0).reset_index(
+        drop=True,
+    )
+    assert len(df_by_eval) == 1
+    assert eval_cfg is None
+    _assert_frame_equal(df_datapoints, expected_df_dp, dp_columns)
+    _assert_frame_equal(df_inferences, expected_df_inf, inf_columns)
+    _assert_frame_equal(df_metrics, expected_df_mtr, mtr_columns)
+
+
+def test__test__invalid_data__eval_before_inf() -> None:
+    dataset_name = with_test_prefix(f"{__file__}::test__test__invalid_data__eval_before_inf")
+    model_name = with_test_prefix(f"{__file__}::test__test__invalid_data__eval_before_inf")
+    df_dp = get_df_dp()
+    dp_columns = ["user_dp_id", "locator", "width", "height", "city"]
+    register_dataset(dataset_name, df_dp[13:20][dp_columns])
+
+    df_mtr = get_df_mtr()
+    mtr_columns = ["score"]
+
+    with pytest.raises(RemoteError) as exc_info:
+        test(
+            dataset_name,
+            model_name,
+            eval=get_eval_func(df_mtr, mtr_columns),
+        )
+    exc_info_value = str(exc_info.value)
+    assert "invalid input data" in exc_info_value
+
+
+def test__test__invalid_data__df_size_mismatch() -> None:
+    dataset_name = with_test_prefix(f"{__file__}::test__test__invalid_data__df_size_mismatch")
+    model_name = with_test_prefix(f"{__file__}::test__test__invalid_data__df_size_mismatch")
+    df_dp = get_df_dp()
+    dp_columns = ["user_dp_id", "locator", "width", "height", "city"]
+    register_dataset(dataset_name, df_dp[dp_columns])
+
+    df_inf = get_df_inf(10)
+    df_mtr = get_df_mtr(10)
+    inf_columns = ["softmax_bitmap"]
+    mtr_columns = ["score"]
+
+    with pytest.raises(IncorrectUsageError) as exc_info:
+        test(
```

Comment:
Are we able to catch missed inference/metrics and inform users before relying on server side to return `ServiceDataIngestion` error due to null check? That error is harder to understand imho.

On another note, is there legit cases to allow partial results? cc @y27choi

Reply:
i don't think sdk-server side will raise `ServiceDataIngestion` due to null check b/c we will always have a dict for the inf or metrics as long as the df length is identical with df_dp.
see test `test__test__upload_none`, and we will save `{"softmax_bitmap": null }` in the db

Reply:
Sorry, commented on the wrong line, I was referring to line #294 in test case `test__test__invalid_data__eval_before_inf`. A similar situation would be if `test` invoked after line #230 in test case `test__test__missing_metric`.

Reply:
just add the strict client logic to catch this and fail.
based on our [sync](https://kolena-io.slack.com/archives/C01GAKM1W5S/p1697045451173549?thread_ts=1696983622.468879&cid=C01GAKM1W5S), we might want to allow the partial results. technically we just drop all invalid rows before uploading and have some warning msg.

or we could add a parameter in the `test` func to control this behaviour, but i am a bit hesitate for it due to more complex public interface